_posts/2024-04-04-cheriot-supply-chain.markdown

```diff
@@ -0,0 +1,189 @@
+The paper noted that, with Capsicum on conventional hardware, it was easy to sandbox the `gzip` program that wrapped zlib, but not easy to sandbox the library.
+You may notice two things about that paper:
+
+ - One of the authors was a PI on the original CHERI research grant, another was on the advisory board.
+ - It was published the same year that CHERI development started.
+
+Neither of these is a coincidence.
+CHERI was originally created to address the library-compartmentalisation problem.
+This problem comes from a very simple observation:
+
+**Isolation is easy, sharing is hard.**
+
+If you want to fully isolate two workloads, you can run them on different computers.
+The problems come when you want to share resources.
+When you start looking at libraries, the problem is even more pronounced because the things that you want to share are at the granularity of objects, not files or pages.
+
+CHERI was designed to allow you to load a library with no rights and pass it pointers to objects, which would allow it to access those objects and nothing else.
+I joined the project in 2012 to lead the compilers / language strand of the research.
+Our first approach involved annotating buffers that may be shared so that they could use CHERI capabilities and everything else could use integer addresses to represent pointers.
+It rapidly became apparent that it was not a good strategy because it meant that all code that touched an object had to be aware that the object might be shared.
+Instead, we moved to the 'pure-capability' model, where every pointer was represented as a capability.
+This involved orders of magnitude less porting effort and had the nice benefit of giving fine-grained memory safety for all objects.
+
+It's worth remembering at this point that memory safety was never the goal of CHERI, it's just something that you need to be able to solve the real problems.
+With object-granularity memory safety, you can build a linkage model that lets you link something like libz or liblzma and know that, at worst, it can tamper with the output buffers that you give it.
+On a CHERI system, every pointer in the source language is lowered to a *CHERI capability*, an unforgeable hardware type that authorises access to memory and carries bounds and permissions.
+You can think of a CHERI capability as a hardware-enforced pointer.
+Every memory access (load, store, or jump) instruction takes a capability that authorises the operation as an operand.
+You can shrink the bounds or remove permissions to create new pointers with restricted rights, and that's a simple register-to-register operation.
+
+This is where the CHERIoT project started.
+We wanted to build a compartmentalisation model (and an associated programmer model) that can rely on CHERI existing and use this to build the kind of supply-chain security where you can fearlessly use third-party code.
```

Comment:
```suggestion
The CHERIoT project builds atop this background.
We wanted to build a compartmentalisation model (and an associated programmer model)  for embedded systems that would allow us to address these supply-chain security concerns and fearlessly use third-party code.
The only presently viable foundational technology is CHERI, and so CHERIoT was born.
```